Enhance shared library support for AIX

KamathForAIX opened the pull request:
This PR is for fixing issue number #12219 in meson for AIX operating system as discussed.

KamathForAIX commented:
Respected @eli-schwartz and community members, 

Kindly review this pull request. Let me know what you think.

eli-schwartz commented:
I'm putting this on my schedule to review on Sunday. :)

KamathForAIX commented:
Thank you.

eli-schwartz commented on mesonbuild/modules/python.py:
```diff
@@ -232,6 +232,11 @@ def extension_module_method(self, args: T.Tuple[str, T.List[BuildTargetSource]],
         kwargs['name_prefix'] = ''
         kwargs['name_suffix'] = target_suffix
 
+        # In AIX if we are using a extension module, this means a python module does not
+        # require a archived shared library due to dlopen mechanism of python. Hence do not archive.
+        if mesonlib.is_aix():
+            SharedLibrary.aix_so_archive = False
```
We should avoid mutating global state like this.

eli-schwartz replied:
I think we can just get rid of it altogether because we:
```python
        return self.interpreter.build_target(self.current_node, args, kwargs, SharedModule)
```

So we are already handling a SharedModule, and we *should* already be getting skipped due to the condition:
```python
if isinstance(target, build.SharedLibrary) and not isinstance(target, build.SharedModule)
```

KamathForAIX replied:
You are right here. Yes, we should remove this.

eli-schwartz commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1057,9 +1057,10 @@ def generate_target(self, target):
-            elem = NinjaBuildElement(self.all_outputs, linker.get_archive_name(outname), 'AIX_LINKER', [outname])
-            self.add_build(elem)
+        if isinstance(target, build.SharedLibrary) and not isinstance(target, build.SharedModule) and self.environment.machines[target.for_machine].is_aix():
+            if target.aix_so_archive is True:
```
What if we set `aix_so_archive = False` as a property on SharedModule, to override the one on SharedLibrary?

KamathForAIX replied:
Thank you @eli-schwartz for your review. 

This is actually better. So if we set aix_so_archive variable to False in SharedModule and override it, we will not have to do any checks in ninjabackend.py for SharedModule at all. It will just override it and when py3.extension_module is called, aix_so_archive will be False there, as a consequence we will not need to archive. 

Amazing!! Thanks once again @eli-schwartz for your guidance. When I checked this with glib2, scipy and p11 kit, this logic worked. I will soon commit again with these changes. 
 
 Have a nice day ahead.

KamathForAIX commented:
Respected @eli-schwartz and community members,

I addressed the comments mentioned in the review. Kindly let me know what you think now and if any further changes needed from my end. 

While I tried to add conditions for AIX code so that no other targets are affected I still see one failure in checks which I could not figure out why my changes have caused the same. Since you all are experts, Kindly let me know what can be the root cause of the same. 

Have a nice day ahead.

eli-schwartz commented on mesonbuild/backend/backends.py:
```diff
@@ -355,6 +355,10 @@ def get_target_filename_for_linking(self, target: T.Union[build.Target, build.Cu
         if isinstance(target, build.SharedLibrary):
             link_lib = target.get_import_filename() or target.get_filename()
+            # In AIX, if we archive .so, the blibpath must link to archived shared library otherwise to the .so file.
+            if mesonlib.is_aix():
+                if target.aix_so_archive is True:
```
`if target.aix_so_archive:`

(You don't need "is True" here.)

I would also put it all on a single line:
```python
if mesonlib.is_aix() and target.aix_so_archive:
```

KamathForAIX replied:
Done

eli-schwartz commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -1057,9 +1057,10 @@ def generate_target(self, target):
-        if isinstance(target, build.SharedLibrary) and self.environment.machines[target.for_machine].is_aix():
-            elem = NinjaBuildElement(self.all_outputs, linker.get_archive_name(outname), 'AIX_LINKER', [outname])
-            self.add_build(elem)
+        if self.environment.machines[target.for_machine].is_aix() and isinstance(target, build.SharedLibrary):
```
Random question: why is the order being reversed here? Was there something wrong with doing `isinstance(...) and self.environment.*` that this needed to be switched?

KamathForAIX replied:
> Random question: why is the order being reversed here? Was there something wrong with doing `isinstance(...) and self.environment.*` that this needed to be switched?

There was nothing wrong. I was just trying to see if that msys2 clang build failure had got to do something with it. So I was using the short circuit method of python to evaluate. So if the platform is not AIX, then it will not evaluate the second part of the condition and after that if that msys2 clang build failure passed (Though in reality I knew it does not matter) then I might had to think why that happened. 

So yeah, that is why I tried that.

eli-schwartz replied:
Then can you please restore it to its original state? :)

(And squash the Pull Request into one commit per mergeable, logical, change. In this case the entire PR should be one logical change -- changing the logic to use the .target_aix_so_archive property everywhere.)

KamathForAIX replied:
@eli-schwartz Here we go. Done :)

Thanks..

eli-schwartz commented:
> While I tried to add conditions for AIX code so that no other targets are affected I still see one failure in checks which I could not figure out why my changes have caused the same. You all are experts and have immense knowledge of meson, Kindly let me know what can be the root cause of the same with respect to these changes.

This failing test (msys2 clang handling of LLVM) is failing everywhere, not just on your PR. Don't worry about it.

KamathForAIX commented:
> > While I tried to add conditions for AIX code so that no other targets are affected I still see one failure in checks which I could not figure out why my changes have caused the same. You all are experts and have immense knowledge of meson, Kindly let me know what can be the root cause of the same with respect to these changes.
> 
> This failing test (msys2 clang handling of LLVM) is failing everywhere, not just on your PR. Don't worry about it.

Thank you so much for this confirmation @eli-schwartz. 

So I made the suggested changes. Kindly let me know if any more changes are needed. 

Have a nice day ahead. 

Thanks and regards,
Aditya.

KamathForAIX commented:
Respected @eli-schwartz and community members,

Kindly let me know if I missed something in the review comments. I addressed the review changes suggested and squashed the commit into one message. 

Kindly merge this PR if there are no changes needed. Thank you for your guidance. 

Regards,
Aditya Kamath.

eli-schwartz commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -2381,13 +2382,13 @@ def generate_dynamic_link_rules(self):
-                cmdlist = compiler.get_command_to_archive_shlib()
-                args = []
-                options = {}
-                self.add_rule(NinjaRule(rule, cmdlist, args, description, **options, extra=None))
+        if self.environment.machines[for_machine].is_aix():
+            rule = 'AIX_LINKER{}'.format(self.get_rule_suffix(for_machine))
+            description = 'Archiving AIX shared library'
+            cmdlist = compiler.get_command_to_archive_shlib()
+            args = []
+            options = {}
+            self.add_rule(NinjaRule(rule, cmdlist, args, description, **options, extra=None))
```
Why is this indentation being removed here? It looks like you are dropping off the end of the for loop and only producing one rule, using whichever for_machine turns out to be last...

KamathForAIX replied:
@eli-schwartz 

You are right here. I changed it back. Also, while I changed it back, I saw an error while building Scipy, which looked like this:

```
File "/meson_install/opt/freeware/lib/python3.9/site-packages/mesonbuild/backend/ninjabackend.py", line 1315, in generate_rules
    self.generate_dynamic_link_rules()
  File "/meson_install/opt/freeware/lib/python3.9/site-packages/mesonbuild/backend/ninjabackend.py", line 2388, in generate_dynamic_link_rules
    cmdlist = compiler.get_command_to_archive_shlib()
  File "/meson_install/opt/freeware/lib/python3.9/site-packages/mesonbuild/compilers/compilers.py", line 963, in get_command_to_archive_shlib
    return self.linker.get_command_to_archive_shlib()
AttributeError: 'NoneType' object has no attribute 'get_command_to_archive_shlib'
```
I added a guard in the compilers.py file. Kindly see the changes there and let me know. This is because linker object was none.

eli-schwartz commented on mesonbuild/backend/ninjabackend.py:
```diff
@@ -3699,8 +3695,8 @@ def generate_ending(self) -> None:
                 #Add archive file if shared library in AIX for build all.
-                if isinstance(t, build.SharedLibrary):
-                    if self.environment.machines[t.for_machine].is_aix():
+                if self.environment.machines[t.for_machine].is_aix():
+                    if isinstance(t, build.SharedLibrary) and t.aix_so_archive:
```
Can you flip this around? it should be the same except matching the original order of conditions:

```suggestion
                if isinstance(t, build.SharedLibrary) and t.aix_so_archive:
                    if self.environment.machines[t.for_machine].is_aix():
```

KamathForAIX replied:
Done

eli-schwartz commented:
Apologies for the lateness of the reply, there were some IRL things going on (all good news, I promise! :smile:)

KamathForAIX commented:
> Apologies for the lateness of the reply, there were some IRL things going on (all good news, I promise! 😄)

Thank you for your reply. No problem. Please let me know now. I have made those changes.

eli-schwartz commented:
I have tweaked the commit messages and moved some code in the second commit back into the first commit where it seemed to belong. Here is a range-diff of the changes:
```diff
$ git --no-pager range-diff origin/master 1563bac5e cfec25502
1:  2bac26601 ! 1:  f4d19db25 Use target_aix_so_archive to decide to archive shared library in AIX
    @@ Metadata
     Author: Aditya Vidyadhar Kamath <Aditya.Kamath1@ibm.com>
     
      ## Commit message ##
    -    Use target_aix_so_archive to decide to archive shared library in AIX
    +    Use target.aix_so_archive to decide to archive shared library in AIX
     
    -    Enhance shared library support for AIX
    +    Previously, AIX support was updated to archive shared libraries per AIX
    +    platform conventions, which expect .a files that contain .so files. This
    +    is usually correct, but an edge case occurs for loadable plugins, e.g.
    +    what meson creates for `shared_module()`. A notable example is python
    +    extensions (SciPy, for example).
     
    -    Change == to is and simplify patch to fix build break
    +    These should *not* be archived, because the .so file itself needs to be
    +    loaded as a plugin. For example, SciPy fails to import in the python
    +    interpreter.
     
    -    Simplify the logic by overriding aix_so_archive in SharedModule
    +    Handle this by differentiating between plugins and regular libraries,
    +    and only archiving when safe to do so.
     
    -    Check if AIX is the machine always
    -
    -    Fix to cross compile build break
    -
    -    Avoid using aix_so_archive in other targets
    -
    -    Remove unnecessary conditions
    -
    -    Add comment to condition change in install section for AIX
    -
    -    Simply if condition to a simple and logic
    -
    -    Use only target.aix_so_archive and change one cond to original
    +    Fixes #12219
     
      ## mesonbuild/backend/backends.py ##
     @@ mesonbuild/backend/backends.py: class Backend:
    @@ mesonbuild/backend/ninjabackend.py: class NinjaBackend(backends.Backend):
      
          def should_use_dyndeps_for_target(self, target: 'build.BuildTarget') -> bool:
              if mesonlib.version_compare(self.ninja_version, '<1.10.0'):
    -@@ mesonbuild/backend/ninjabackend.py: class NinjaBackend(backends.Backend):
    - 
    -                 options = self._rsp_options(compiler)
    -                 self.add_rule(NinjaRule(rule, command, args, description, **options, extra=pool))
    --            if self.environment.machines[for_machine].is_aix():
    --                rule = 'AIX_LINKER{}'.format(self.get_rule_suffix(for_machine))
    --                description = 'Archiving AIX shared library'
    --                cmdlist = compiler.get_command_to_archive_shlib()
    --                args = []
    --                options = {}
    --                self.add_rule(NinjaRule(rule, cmdlist, args, description, **options, extra=None))
    -+        if self.environment.machines[for_machine].is_aix():
    -+            rule = 'AIX_LINKER{}'.format(self.get_rule_suffix(for_machine))
    -+            description = 'Archiving AIX shared library'
    -+            cmdlist = compiler.get_command_to_archive_shlib()
    -+            args = []
    -+            options = {}
    -+            self.add_rule(NinjaRule(rule, cmdlist, args, description, **options, extra=None))
    - 
    -         args = self.environment.get_build_command() + \
    -             ['--internal',
     @@ mesonbuild/backend/ninjabackend.py: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47485'''))
              else:
                  dependencies = target.get_dependencies()
    @@ mesonbuild/backend/ninjabackend.py: https://gcc.gnu.org/bugzilla/show_bug.cgi?id
                      # they are all built
                      #Add archive file if shared library in AIX for build all.
     -                if isinstance(t, build.SharedLibrary):
    --                    if self.environment.machines[t.for_machine].is_aix():
    -+                if self.environment.machines[t.for_machine].is_aix():
    -+                    if isinstance(t, build.SharedLibrary) and t.aix_so_archive:
    ++                if isinstance(t, build.SharedLibrary) and t.aix_so_archive:
    +                     if self.environment.machines[t.for_machine].is_aix():
                              linker, stdlib_args = self.determine_linker_and_stdlib_args(t)
                              t.get_outputs()[0] = linker.get_archive_name(t.get_outputs()[0])
    -                 targetlist.append(os.path.join(self.get_target_dir(t), t.get_outputs()[0]))
     
      ## mesonbuild/build.py ##
     @@ mesonbuild/build.py: class SharedLibrary(BuildTarget):
2:  1563bac5e < -:  --------- Fix to 'NoneType' object has no attribute 'get_command_to_archive_shlib'
-:  --------- > 2:  cfec25502 Fix traceback on AIX in shlib archiving code
```

KamathForAIX commented:
Thank you @eli-schwartz for the merge and your guidance. I learned a lot from you. Thank you, community.
